ext/opcache/Optimizer/zend_optimizer.c

```diff
@@ -340,6 +340,23 @@ int zend_optimizer_update_op1_const(zend_op_array *op_array,
 		case ZEND_FETCH_LIST_R:
 		case ZEND_COPY_TMP:
 			return 0;
+		case ZEND_ECHO:
+		{
+			zval zv;
+			if (Z_TYPE_P(val) != IS_STRING && zend_optimizer_eval_cast(&zv, IS_STRING, val) == SUCCESS) {
+				opline->op1.constant = zend_optimizer_add_literal(op_array, &zv);
+				val = CT_CONSTANT_EX(op_array, opline->op1.constant);
+			} else {
+				opline->op1.constant = zend_optimizer_add_literal(op_array, val);
+			}
```

Comment:
I'd write this as:

```
			if (Z_TYPE_P(val) != IS_STRING && zend_optimizer_eval_cast(&zv, IS_STRING, val) == SUCCESS) {
				zval_ptr_dtor_nogc(val);
				val = &zv;
			}
			opline->op1.constant = zend_optimizer_add_literal(op_array, val);
```

Destroying `val` is technically not necessary because it so happens that the cast can only succeed if the original value wasn't refcounted, but I think this makes the logic of the code more obvious.

Reply:
Is IS_RESOURCE possible (e.g. STDOUT)? That is refcounted

Reply:
I believe inside opcache it's not possible.